reactive-banana/src/Reactive/Banana/Prim/Plumbing.hs

```diff
@@ -129,20 +130,21 @@ addOutput p = do
     Build monad
 ------------------------------------------------------------------------------}
 runBuildIO :: BuildR -> BuildIO a -> IO (a, Action, [Output])
-runBuildIO i m = do
-        (a, BuildW (topologyUpdates, os, liftIOLaters, _)) <- unfold mempty m
+runBuildIO !i m = do
+        (a, BuildW (topologyUpdates, os, liftIOLaters, _)) <- unfold i mempty m
         doit liftIOLaters          -- execute late IOs
         return (a,Action $ Deps.buildDependencies topologyUpdates,os)
-    where
-    -- Recursively execute the  buildLater  calls.
-    unfold :: BuildW -> BuildIO a -> IO (a, BuildW)
-    unfold w m = do
-        (a, BuildW (w1, w2, w3, later)) <- RW.runReaderWriterIOT m i
-        let w' = w <> BuildW (w1,w2,w3,mempty)
-        w'' <- case later of
-            Just m  -> snd <$> unfold w' m
-            Nothing -> return w'
-        return (a,w'')
+{-# inline runBuildIO #-}
+
+-- Recursively execute the  buildLater  calls.
+unfold :: BuildR -> BuildW -> BuildIO a -> IO (a, BuildW)
+unfold !i w m = do
+    (a, BuildW (w1, w2, w3, later)) <- RW.runReaderWriterIOT m i
+    let !w' = w <> BuildW (w1,w2,w3,mempty)
+    w'' <- case later of
+        Just m  -> snd <$> unfold i w' m
+        Nothing -> return w'
+    return (a,w'')
```

Comment:
I don't necessarily need to make `unfold` a top-level binding, but the important thing is that it doesn't mention any variables bound by `runBuildIO`. This allows us to inline a bit of `runBuildIO`, and to float the recursion loop out to the top-level. Otherwise, each `runBuildIO` allocates a new recursion closure